src/internal/packager2/pull.go

```diff
@@ -0,0 +1,228 @@
+
+	switch u.Scheme {
+	case "oci":
+		err := pullOCI(ctx, src, tmpPath, shasum, filter)
+		if err != nil {
+			return err
+		}
+	case "http", "https":
+		err := pullHTTP(ctx, src, tmpPath, shasum)
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("unknown scheme %s", u.Scheme)
+	}
+
+	name, err := nameFromMetadata(tmpPath)
+	if err != nil {
+		return err
+	}
+	tarPath := filepath.Join(dir, name)
+	err = os.Remove(tarPath)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return err
+	}
+	dstFile, err := os.Create(tarPath)
+	if err != nil {
+		return err
+	}
+	defer dstFile.Close()
+	srcFile, err := os.Open(tmpPath)
+	if err != nil {
+		return err
+	}
+	defer srcFile.Close()
+	_, err = io.Copy(dstFile, srcFile)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func pullOCI(ctx context.Context, src, tarPath, shasum string, filter filters.ComponentFilterStrategy) error {
+	tmpDir, err := utils.MakeTempDir(config.CommonOptions.TempDirectory)
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmpDir)
+	if shasum != "" {
+		src = fmt.Sprintf("%s@sha256:%s", src, shasum)
+	}
+	arch := config.GetArch()
+	remote, err := zoci.NewRemote(src, oci.PlatformForArch(arch))
+	if err != nil {
+		return err
+	}
+	desc, err := remote.ResolveRoot(ctx)
+	if err != nil {
+		return fmt.Errorf("could not fetch images index: %w", err)
+	}
+	layersToPull := []ocispec.Descriptor{}
+	if supportsFiltering(desc.Platform) {
+		pkg, err := remote.FetchZarfYAML(ctx)
+		if err != nil {
+			return err
+		}
+		pkg.Components, err = filter.Apply(pkg)
+		if err != nil {
+			return err
+		}
+		layersToPull, err = remote.LayersFromRequestedComponents(ctx, pkg.Components)
+		if err != nil {
+			return err
+		}
+	}
+	_, err = remote.PullPackage(ctx, tmpDir, config.CommonOptions.OCIConcurrency, layersToPull...)
+	if err != nil {
+		return err
+	}
+	allTheLayers, err := filepath.Glob(filepath.Join(tmpDir, "*"))
+	if err != nil {
+		return err
+	}
+	err = archiver.Archive(allTheLayers, tarPath)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func pullHTTP(ctx context.Context, src, tarPath, shasum string) error {
+	f, err := os.Create(tarPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, src, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		_, err := io.Copy(io.Discard, resp.Body)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("unexpected http response status code %s for source %s", resp.Status, src)
+	}
+	_, err = io.Copy(f, resp.Body)
+	if err != nil {
+		return err
+	}
+	received, err := helpers.GetSHA256OfFile(tarPath)
+	if err != nil {
+		return err
+	}
+	if received != shasum {
+		return fmt.Errorf("shasum mismatch for file %s, expected %s but got %s", tarPath, shasum, received)
```

Comment:
I think this error message looks confusing. Could we instead give the user the url rather than the file. Additionally, I think we should error at the start of this function if shasum == ""

```
shasum mismatch for file /tmp/zarf-3623879537/data.tar.zst, expected but got
             690799dbe8414238e11d4488754eee52ec264c1584cd0265e3b91e3e251e8b1a
```